### PR TITLE
Implementation of Morse Potential

### PR DESCRIPTION
**Pull request by EloD10**
I try to implement Morse Potential correctly =)

Thanks to Luthaf for help !

### CODE REVIEW

File: src/core/src/energy/functions.rs [hunk 1]
```diff
@@ -299,6 +299,34 @@ impl PairPotential for Buckingham {
     }
 }
 
+struct MorsePotential {
+    with: f64,
```
Comment (Luthaf) [inline src/core/src/energy/functions.rs]: I think this is a typo: `with` => `width`

File: src/core/src/energy/functions.rs [hunk 2]
```diff
@@ -299,6 +299,34 @@ impl PairPotential for Buckingham {
...
 
+struct MorsePotential {
+    with: f64, 
+    r0: f64,
+    depth: f64 
+}
+
+impl Potential for MorsePotential {  
+
+fn energy(&self, r: f64) -> f64 {
```
Comment (Luthaf) [inline src/core/src/energy/functions.rs]: Could you please indent the function like the others one? 

```
impl Potential for MorsePotential {
    fn energy(&self, r: f64) -> f64 {
        // Code goes here
    }
}
```

**Comment by Luthaf** [comment]
Thank you for taking the time to send a PR!

Travis testing is [failling](https://travis-ci.org/lumol-org/lumol/jobs/246160946#L433) with various errors, can you make sure the code compiles locally?

PS: feel free to ask for a french translation of any comment here!

File: src/core/src/energy/functions.rs [hunk 3]
```diff
@@ -299,6 +299,34 @@ impl PairPotential for Buckingham {
...
+impl Potential for MorsePotential {  
+
+fn energy(&self, r: f64) -> f64 { 
+    let rc2 = (1.0 - f64::exp(self.with * (r - self.r0))) * (1.0 - f64::exp(self.with * (r - self.r0)));
+    let result = self.depth * rc2;
+    return result;
+    }
+
+fn force(&self, r: f64) -> f64 { 
+    let result = self.depth * (std::exp(-2 * self.with * (r - self.r0)) - 2 * std::exp(self.with * (r -self.r0))) ;
```
Comment (Luthaf) [inline src/core/src/energy/functions.rs]: `std::exp` should be `f64::exp`.

File: src/core/src/energy/functions.rs [hunk 4]
```diff
@@ -299,6 +299,34 @@ impl PairPotential for Buckingham {
     }
 }
 
+struct MorsePotential {
```
Comment (Luthaf) [inline src/core/src/energy/functions.rs]: The struct is missing documentation. You can take inspiration from the other potentials in this file.

Comment (EloD10) [inline src/core/src/energy/functions.rs]: I will try to do this soon !

File: src/core/src/energy/functions.rs [hunk 5]
```diff
@@ -364,6 +363,33 @@ impl PairPotential for BornMayerHuggins {
...
+impl Potential for MorsePotential {  
+    fn energy(&self, r: f64) -> f64 { 
+        let rc2 = (1.0 - f64::exp(self.width * (r - self.r0))) * (1.0 - f64::exp(self.width * (r - self.r0)));
+        let result = self.depth * rc2;
+        return result;
+    }
+    
+    fn force(&self, r: f64) -> f64 { 
+        let result = self.depth * (f64::exp(-2 * self.width * (r - self.r0)) - 2 * f64::exp(self.width * (r -self.r0))) ; 
+        return result;
```
Comment (antoinewdg) [inline src/core/src/energy/functions.rs]: Just so you know, [you do not need the return statement](https://doc.rust-lang.org/book/second-edition/ch03-03-how-functions-work.html#functions-with-return-values) in Rust. This is not a big issue, but in this case would make 
the functions simpler :)

**Comment by g-bauer** [comment]
Hey, thanks for the PR! I never used this potential, is it used to describe inter- or intramolecular energies? Do you have a specific use case in mind?

**Comment by Luthaf** [comment]
> I never used this potential, is it used to describe inter- or intramolecular energies? Do you have a specific use case in mind?

I never used it either, but it is described as a better bonding potential, for bonds, angles and dihedrals (see the wikipedia page: https://en.wikipedia.org/wiki/Morse_potential). It improves on harmonic potential by being breakable, and by representing anharmonicity.

Also, it is an easy task for getting started with rust =)

**Comment by Luthaf** [comment]
Closing in favor of #192

**Pull request closed without merging**